### ISSUE TITLE
Support Elixir durations

### ISSUE THREAD

**Issue by josevalim** [elixir-ecto/postgrex#688]
### Elixir version

1.17.0

### Database and Version

Any

### Postgrex Version

Not yet supported

### Current behavior

Postgrex.Duration is not supported.

### Expected behavior

We should support the new durations and potentially deprecate Postgrex.Duration.

**Comment by greg-rychlewski** [comment]
For now were you just looking to encode Elixir durations? And then when we require 1.17 we can do the decoding/replace Postgrex.Duration?

**Comment by josevalim** [comment]
I think that's a good beginning, yeah! We could maybe already even allow defaulting to them via an option?

**Comment by josevalim** [comment]
This is in but we want to keep it around and eventually make it the default.

**Comment by endoooo** [comment]
after adjusting the config following the [documentation](https://hexdocs.pm/postgrex/readme.html#:~:text=(3)%20%25Duration%7B%7D%20may%20only%20be%20used%20with%20Elixir%201.17%2B) (thanks to the tips in #706 and [this Elixir Forum topic](https://elixirforum.com/t/difficulty-employing-duration-to-postgresql-interval-with-solution/67990)) everything was **apparently** working just fine.

I said apparently because while preparing a release, I noticed a strange behavior:

| inspect in UI | query in DB |
| - | - |
| <img width="609" alt="Image" src="https://github.com/user-attachments/assets/5cf409b3-453e-42ae-82f1-e962d72d35e7" /> | <img width="481" alt="Image" src="https://github.com/user-attachments/assets/571ae828-1e1d-4e4f-b03a-0732cf477d2c" /> |

the DB interval is not matching the `Duration` in the application.

maybe this is somehow related to the notes described [here](https://github.com/elixir-ecto/postgrex/pull/696#issue-2431418697)?

---

relevant code

```elixir
# config
config :my_app, MyApp.Repo, types: MyApp.PostgrexTypes

# my_app/postgrex_types.ex
Postgrex.Types.define(MyApp.PostgrexTypes, Ecto.Adapters.Postgres.extensions(),
  interval_decode_type: Duration
)

# migration
alter table(:table_name) do
  #...
  add :duration, :duration,
    generated: """
      ALWAYS AS (CASE
        WHEN closed_at IS NOT NULL THEN closed_at - inserted_at
        ELSE NULL
      END) STORED
    """
  #...
end

# schema
schema "students_records" do
  #...
  field :duration_until_close, :duration, read_after_writes: true
  #...
end
```

---

ps: sorry if this is not the right place to post it.

ps2: if someone can hint me with a starting point, I can try to investigate it and contribute with a PR (although, as a designer, my dev skills are limited haha)

**Comment by greg-rychlewski** [comment]
Thanks @endoooo.

So the issue is that no matter how you insert the duration into Postgres, it will translate it internally into months, days, and microseconds. And that is how you will get the data back out of the database as well.

Given this, the current behaviour is not necessarily a bug but rather one representation choice out of many possible ones. The ones I can think of:

- give the same units Postgres stores (months, days and microseconds) but then this won't be the same as the previous data type, Postgrex.Interval. we probably want some consistency between the two
- make it exactly the same units as Postgrex.Interval (months, days, seconds, microseconds)
- the current behaviour, which is try to fill the largest units first

I can see arguments for all 3....I don't think there is necessarily a clear cut answer. Given Postgrex.Interval has stood for quite a long time that was why I chose to basically do the same thing. 

But let's see what @josevalim thinks.

**Comment by greg-rychlewski** [comment]
Maybe the best compromise is this 

> make it exactly the same units as Postgrex.Interval (months, days, seconds, microseconds)

**Comment by josevalim** [comment]
Yes, I think mirroring Postgrex.Interval is a good point :+1:, and I also think that's the closest representation we can have to PostgreSQL itself, so we get 1 and 2.

**Comment by endoooo** [comment]
ahhhh... now I get it. thank you for the clarification @greg-rychlewski ! I knew it was some misconception on my part 😅 

I was ready to create a PR with @josevalim suggestion, but I noticed you've already done this hehe

thank you guys!
